### PR TITLE
DRAFT: methonds that create the message out of an image or a text

### PR DESCRIPTION
**Pull request by martacerri**

### CODE REVIEW

File: backend/app/api/routes/dde.py [hunk 1]
```diff
@@ -426,48 +429,74 @@ async def extract_from_file(
...
-                ),
-                example.data,
-            )
-            for example in document_data_extractor.document_data_examples
-        )
-    )
-    # Pull data from the file
-    if upload.content_type != "application/pdf":
+        
+    allowed_mime_types = [
```
Comment (lccnl) [inline backend/app/api/routes/dde.py:433]: this should be an Enum ContentType

File: backend/app/handler.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+#two methonds that create the message out of an image or a text 
+
+##add a logic to choose between text or png when i have a pdf 
+    #start extracting information from text and if the logprob are too low use png
+    #add a setting in the dde when i upload examples
+
+##add a logic of looking up the type and converting the input into text or png 
+
+async def full_prompt_from_text(system_prompt:str, prompt:str, examples)-> list[ChatCompletionMessage]:
```
Comment (lccnl) [inline backend/app/handler.py:18]: examples need to be typed

File: backend/app/handler.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        validate_extracted_text(input_text)
+        examples_text += (
+            f"####\nINPUT: {input_text}\n\nOUTPUT: {output_text}\n\n"
+        )
+    full_system_content = f"{system_prompt}\n{examples_text}"
+    messages = [
+        ChatCompletionMessage(role="system", content=full_system_content),
+        ChatCompletionMessage(
+            role="user",
+            content=f"Maintenant, faites la même extraction sur un nouveau document d'input:\n####\nINPUT:{prompt}",
```
Comment (lccnl) [inline backend/app/handler.py:30]: weird to have one message in french and one below in english

File: backend/app/handler.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    ]
+    return messages
+
+def fill_input_image(buffer: BinaryIO) -> dict:
+    img_bytes = buffer.getvalue()
+    base64_image=(base64.b64encode(img_bytes).decode('utf-8'))
+    return {
+        "type": "image_url",
+        "image_url": {
+            "url": f"data:image/png;base64,{base64_image}",
```
Comment (lccnl) [inline backend/app/handler.py:41]: if 'data:image/png;base64' is a constant it should defined on top of the module

File: backend/app/handler.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+    
+#rewrite with ChatCompletionMessage
+#link to the OpenAI documentation specifying that only role 'user' is used for images.
+#https://platform.openai.com/docs/guides/vision
+async def full_prompt_from_image(system_prompt:str, prompt: BinaryIO, examples) -> list[ChatCompletionMessage]:
```
Comment (lccnl) [inline backend/app/handler.py:48]: type examples

File: backend/app/models.py [hunk 6]
```diff
@@ -343,7 +343,8 @@ class DocumentDataExtractorOut(DocumentDataExtractorBase):
     owner_id: int
     document_data_examples: list["DocumentDataExample"]
     response_template: str
-
+    type: str
```
Comment (lccnl) [inline backend/app/models.py:346]: this is a bit vague: I would rather have something as : `process_as`?

File: backend/app/api/routes/dde.py [hunk 7]
```diff
@@ -426,52 +433,23 @@ async def extract_from_file(
...
-            )
-            for example in document_data_extractor.document_data_examples
-        )
-    )
-    # Pull data from the file
-    if upload.content_type != "application/pdf":
+    
+    try:
+        upload_content_type= ContentType(upload.content_type)     
+    except:
```
Comment (lccnl) [inline backend/app/api/routes/dde.py:439]: bare excepts are bad practices: you should catch the correct exception being `ValueError`

File: backend/app/api/routes/dde.py [hunk 8]
```diff
@@ -505,8 +483,7 @@ async def extract_from_file(
     token_indices=map_characters_to_token_indices(extracted_data_token)
     regex_spans=find_value_spans(extracted_data)
     logprobs_sum=get_token_spans_and_logprobs(token_indices, regex_spans, extracted_data_token)
-    return {"extracted_data": json.loads(json_string), "extracted_logprobs":logprobs_sum}
-
+    return {"extracted_data": json.loads(json_string), "extracted_logprobs":logprobs_sum, "identifier": identifier}
```
Comment (lccnl) [inline backend/app/api/routes/dde.py:486]: why do we need the identifier?

File: backend/app/handlers/prompt_for_image.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+from app.models import DocumentDataExtractor
+from app.ops import tup
+from app.ops.documents import as_png
+from app.services.pdf_reader import pdf_reader
+from app.services.png_reader import png_reader
+from app.models import ContentType
+
+BASE64_IMAGE_PREFIX = "data:image/png;base64"
+     
+def fill_input_image(buffer: BinaryIO) -> dict:
```
Comment (lccnl) [inline backend/app/handlers/prompt_for_image.py:15]: you can create a TypedDict with the correct typing here

File: backend/app/handlers/prompt_for_image.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }
+    
+#TODO rewrite with ChatCompletionMessage
+#link to the OpenAI documentation specifying that only role 'user' is used for images:
+#https://platform.openai.com/docs/guides/vision
+async def full_prompt_from_image(file: BinaryIO, document_data_extractor: DocumentDataExtractor, upload_content_type: ContentType) -> list[ChatCompletionMessage]:
+    if upload_content_type == ContentType.PDF:
+        prompt = pdf_reader.as_png(file)
+    if upload_content_type == ContentType.PNG:
```
Comment (lccnl) [inline backend/app/handlers/prompt_for_image.py:31]: eli

File: backend/app/handlers/prompt_for_image.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        prompt = png_reader.as_png(file)
+
```
Comment (lccnl) [inline backend/app/handlers/prompt_for_image.py:33]: else: 
    raise not ImplementedError(f'Content type {upload_content_type} not supported')

File: backend/app/handlers/prompt_for_text.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            status_code=500,
+            detail="The extracted text from the document is empty. Please check if the document is corrupted.",
+        ) 
+
+async def full_prompt_from_text(file: BinaryIO, document_data_extractor: DocumentDataExtractor, upload_content_type: ContentType) -> list[ChatCompletionMessage]:
+    
+    if upload_content_type == ContentType.PDF:
+        prompt = pdf_reader.as_text(file)
+        validate_extracted_text(prompt)
+    if upload_content_type == ContentType.XLSX or upload_content_type == ContentType.XLS:
```
Comment (lccnl) [inline backend/app/handlers/prompt_for_text.py:25]: elif.. else..

File: backend/app/migrations/response_templates.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+                )
+            document_data_extractors = session.exec(statement).all()  # get all the objects from the db 
+            for dde in document_data_extractors:
+                #get the old schema
+                resp_template=dde.response_template
+                new_resp_template=convert_to_json_schema(resp_template) 
+                #update the dde with the new schema
+                dde.response_template=new_resp_template  
+                new_dde={
+                    "response_tempale":new_resp_template
```
Comment (lccnl) [inline backend/app/migrations/response_templates.py:81]: typo

File: backend/app/migrations/response_templates.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+                    select(DocumentDataExtractor)
+                    .order_by(DocumentDataExtractor.name)
+                )
+            document_data_extractors = session.exec(statement).all()  # get all the objects from the db 
+            for dde in document_data_extractors:
+                #get the old schema
+                resp_template=dde.response_template
+                new_resp_template=convert_to_json_schema(resp_template) 
+                #update the dde with the new schema
+                dde.response_template=new_resp_template
```
Comment (lccnl) [inline backend/app/migrations/response_templates.py:79]: I do not think you need this

File: backend/app/handlers/prompt_for_image.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            "text": f"####\nInput of page {idx}",
+            })
+        messages[0]["content"].append(fill_input_image(binary_io))
+
+    messages[0]["content"].append({
+        "type": "text", 
+        "text": f"Expected Output: {output_text}\n\n"
+    })
+
+    messages[0]["content"].extend([
```
Comment (ngrislain) [inline backend/app/handlers/prompt_for_image.py:81]: Here the prompt is a list (according to the signeture of as_png for pdfs...

File: backend/app/handlers/prompt_for_image.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+            "url": f"{BASE64_IMAGE_PREFIX},{base64_image}",
+        }
+    }
+    
+#TODO rewrite with ChatCompletionMessage
+#link to the OpenAI documentation specifying that only role 'user' is used for images:
+#https://platform.openai.com/docs/guides/vision
+async def full_prompt_from_image(file: BinaryIO, document_data_extractor: DocumentDataExtractor, upload_content_type: ContentType) -> list[ChatCompletionMessage]:
+    if upload_content_type == ContentType.PDF:
+        prompt = pdf_reader.as_png(file)
```
Comment (ngrislain) [inline backend/app/handlers/prompt_for_image.py:37]: Super dangerous to return a BinaryIO in one case and list in the other

**Review by ngrislain — APPROVED** [review]

**Pull request merged**